Initial design document

tcsc opened the pull request:
The first cut of the design document.

tcsc committed:
Initial design document

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,175 @@
+
+### Placeholder types
+
+`task_handle`: A identifier for a running task, unique across the set of currently registered tasks. The exact type this will map to is not yet defined, but likely to be a either a `string` or an `int64`. The data it encodes will *not* be a simple unix PID, as the API server will have to track the state of processes that have exited, and the system may recycle the PIDs in the lifetime of the API  ...
```
What is the algorithm that's gonna be used to generate task ids? Also, the information about data types is not needed in the design doc (to keep it clean and clear from unnecessary details).

tcsc replied:
For the purposes of this exercise, perhaps generating a UUID would work? I can imagine adding some sort of clash-detection & retry mechanism to the task registration system to handle the remote possibility of a clash.

Will remove the discussion of concrete types in an upcoming commit.

tcsc replied:
Done

r0mant replied:
Yep, uuids are fine, let's add this to the design doc.

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,175 @@
+
+### Stopping/Signalling a task
+
+`signal_task(task task_handle, signal int) error`
```
Nit: `signal_task` name is a bit inconsistent with the other 2 methods that use `camelCase`.

tcsc replied:
Too much python... *sigh*. Will change in upcoming commit.

tcsc replied:
Removed as part of extracting API discussion to `proto` file

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,175 @@
+
+### Querying a task
+
+`queryTask(task: task_handle, stdout_range: *range, stderr_range *range) (task_status, []byte, []byte, error)`
```
I would document these APIs in terms of gRPC service/messages that will be used rather than pseudocode.

russjones replied:
Yes, instead just share the proto file with us.

tcsc replied:
Done

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,175 @@
+
+### Freeing task resources
+
+`free_task(task task_handle) error`
```
Same nit as above, `free_task` vs `freeTask`.

Also, do you mean for this to be a part of gRPC API provided by the server? What is the purpose of it, e.g. when would you want to send a TERM signal vs calling this?

russjones replied:
Doesn't seem like this should be exposed to the client. Same with signal_task. The RPC should just be to top, how the server implements it should be internal to the server and client shouldn't have to know about it.

tcsc replied:
I suppose it comes down to what we want the lifecycle of a task to be. The last time I did something like this we had to stream the logs piecemeal to the client, and decide based on the log content when & how to signal the task. I suspect that part of me is re-writing that system. 

Anyway, I propose that the lifecycle of a task should probably look like this:

1. Task is started by a call from the client
2. Client repeatedly polls server for task completion
3. Client may elect to kill task at any time, server will try to shut it down gracefully at first and then brutally after some (server-specified) timeout
4. The client may fetch the logs at any time after starting the task, but will always receive the entire log; no ranges 
5. After some inactivity timeout, the task record on the server is deleted. This implies that the task, if still running, is killed and the logs are no longer retrievable.

If that seems reasonable, I will update the API to:
- remove the `free` endpoint,
- rename the `signal` endpoint to something like `kill` or `quit` & remove the signal number argument, and 
- add a new `fetchLogs` endpoint to return `stdout` and `stderr`

r0mant commented on doc/design.md:
> server will use a signed certificate

Signed by what?

tcsc replied:
Updated discussion to talk about CA requirements, rather than a "signed certificate"

r0mant commented on doc/design.md:
What kind of data will be stored in this config file?

tcsc replied:
I don't actually know yet - because at this point I don't know enough about how gRPC authentication works to know exactly what info I will need.  For an exercise of this scope, "what I will need" may well be _nothing_, obviating the need for a user database or config file altogether.

I just wanted to be up-front that I will not be attempting to dynamically reload the server config during runtime.

I will re-word this paragraph to make the intent clearer.

tcsc replied:
Done (hopefully)

r0mant commented on doc/design.md:
Nit: Extra ")".

tcsc replied:
Removed

r0mant commented on doc/design.md:
Yeah, config files and env vars are out of scope for sure.

russjones commented on doc/design.md:
```diff
@@ -0,0 +1,175 @@
+
+**Arguments**
+-  `task`: The `task_handle` that uniquely identifies a running task. Note that querying a non-existent task, or a task that was not started by the caller, is considered an error.
+-  `stdout_range`: An optional `range` indicating the range of bytes from the task's stdout stream that the caller wants. A `nil` value indicates that the caller is not interested in receiving the logs with this call.
+-  `stderr_range`: An optional `range` indicating the range of bytes from the task's stderr stream that the caller wants. A `nil` value indicates that the caller is not interested in receiving the logs with this call.
```
I would simply, just get all the logs at once. However, you may want a different RPC endpoint for that.

tcsc replied:
Fair enough. Will refactor that in an upcoming commit.

tcsc replied:
Done

fspmarshall reviewed (approved):
LGTM once the existing feedback is addressed.

tcsc committed:
Addressing review concerns:

 - Moves API discussion to `api/levity.proto`
 - Adds detail on task handle generation algorithm
 - Adds discussion on task lifecycle
 - Removes notion of `free`ing a task.
 - Attempts to clarify that server will only read config on startup
 - Attempts to clarify mutual TLS certificate CA requirements
 - Fixes minor typos

russjones commented on doc/design.md:
```diff
@@ -0,0 +1,128 @@
+4. The client may fetch the logs with `FetchLogs` at any time after
+   starting the task, but will always receive the entire log at the time
+   of the call. There is no mechanism to retreive a partial log.
+5. After some inactivity timeout, the task record on the server is deleted.
```
Good idea, but not a requirement, you don't have to implement this.

russjones commented on doc/design.md:
```diff
@@ -0,0 +1,128 @@
+## Task Lifecycle
+
+1. Task is started by a call to `StartTask`.
+2. The client repeatedly polls server for task completion via `QueryTask`.
```
Can you clarify? This is going to be an option for the client to do, not something the client will be doing normally right?

Basically you plan to implement `./client status uuid` that will return right away with the status but the client can be run in a loop if the caller wants with something like `watch`?

tcsc replied:
Essentially yes. Will change the wording to indicate that the user may use the client to do this if they want to monitor a task as it executes.

r0mant commented on api/levity.proto:
```diff
@@ -0,0 +1,91 @@
+
+    // The did not respond to the kill signal and brutaly killed. Implies that 
+    // there There is exit code to return 
+    BrutallyKilled = 3;
```
lol

r0mant commented on api/levity.proto:
```diff
@@ -0,0 +1,91 @@
+    optional int32 exit_code = 2;
+}
+
+message SignalTaskRequest {
```
Should this also include a what kind of signal to send? Or does it only support a single type (like term)?

tcsc replied:
For the purposes of this exrecise, I'm going to say it only supports `SIGTERM`.  This could be added later to the definition of `SignalTaskRequest` if required.

r0mant commented on api/levity.proto:
```diff
@@ -0,0 +1,91 @@
+message StartTaskRequest {
+    string binary = 1;
+    repeated string args = 2;
+    optional string workingDir = 3;
```
Sorry to nitpick to the code that will basically be converted to auto-generated code, but case here is again inconsistent with the rest of the fields that use underscores.

tcsc replied:
Yeah, fair enough. The ProtoBuf style guide says snake case, so may as well fix it now.

tcsc replied:
Done

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,128 @@
+### Transport
+
+The system will use TLS 1.3 for transport level security. The server will
+require a cetificate issued by a CA trusted by any potential clients.
```
certificate

tcsc replied:
Done

r0mant commented on doc/design.md:
```diff
@@ -0,0 +1,128 @@
+The system will use TLS 1.3 for transport level security. The server will
+require a cetificate issued by a CA trusted by any potential clients.
+
+Providing a system for issuing a certificate and having users install the CA root certificate is beyond the scope of this exercise.
```
So how are we going to be able to try it out when it's implemented? I.e. where do we get the CA/keypair for client/server from? Will you provide a script for generating the secrets, or check them into the repo?

tcsc replied:
In order of descending preference:

 1. There will be a makefile target to generate the CA root cert and derive the client and server certs from scratch, or
 2. They will be checked into the repo as independent files, or
 3. They will be baked into the source

Obviously 2 & 3 are not things you would consider doing outside the bounds of an exercise like this. The actual approach will depend on the available time.

tcsc replied:
I've added a brief discussion of this to the _Testing Concerns_ section.

r0mant replied:
Okay, I see, I guess it's just the way it's phrased gave an impression they wouldn't be included in any form.

Yep, the options sounds good, it's ok to check secrets into the repo to keep things simple in this case.

r0mant reviewed (approved):
lgtm when the remaining comments have been addressed.

tcsc committed:
Addresses more review issues

 - Adds CA cert generation comments to "Testing Concerns" section
 - Adds brief discussion of task handle generation algorithm
 - Re-words point 2 of the Task Lifecycle to epmphasise that it's the
   *user* who is resonsible the repeated polling, not the client
   application.
 - Fixes typos

r0mant commented on doc/design.md:
```diff
@@ -86,6 +87,8 @@ This system will use a self-generated CA root certificate for testing purposes.
 
 Ideally, both the client and server will have a configuration option to trust an arbitrary CA in order to run automated tests. Note that this specified CA *may* be hardcoded into the client and server binaries, depending on the time available for implementation. This is obviously something that would not be allowed in a production system.
 
+Time permitting, the project makefile will provide a target to generate both the CD root certificates and derived client & server certificates from scratch.
```
CD -> CA?

tcsc replied:
Oops. Will fix in next MR

r0mant commented on api/levity.proto:
```diff
@@ -65,8 +65,8 @@ enum TaskStatusCode {
 
-    // The did not respond to the kill signal and brutaly killed. Implies that 
-    // there There is exit code to return 
+    // The did not respond to the term signal and brutally killed. Implies that 
```
Nit: "The did not respond" part reads a bit weird.

tcsc replied:
Will fix in next MR

r0mant reviewed (approved):
@tcsc lgtm with just a couple of typos, feel free to merge and proceed to the implementation.
